zng/docs/spec.md

```diff
@@ -350,121 +470,126 @@ previously defined type, the appropriate typedef control code must
+
+| Tag |    Meaning          |
+|-----|---------------------|
+|  0  | nil primitive       |
+|  1  | nil container       |
+|  2  | length 0 primitive  |
+|  3  | length 0 container  |
+|  4  | length 1 primitive  |
+|  5  | length 1 container  |
+|  6  | length 2 primitive  |
+| ... | etc                 |
 
-It is an error for a value to reference a type ID that has not been previously
-defined by a typedef scoped to the stream in which the value appears.
+A container recursively contains a list of tagged values.  Since the container
+encodes its overall length, there is no need to encode the number of elements
+in a container as they are easily discovered by scanning the buffer for each value
+until the last tagged value is encountered.
 
+#### 3.2.2 Tag-Encoded Body of Primitive Values
+
+Following the tag encoding is the value encoded in N bytes as described above.
 A typed value with a `value` of length `N` is interpreted as described in the
-[Primitive Types](#5-primitive-types) table.
+[Primitive Types](#5-primitive-types) table.  The type information needed to
+interpret all of the value elements of a complex type are all implied by the
+top-level type ID of the value message.  For example, the type ID could indicate
+a particular record type, which recursively provides the type information
+for all of the elements within that record, including other complex types
+embedded within the top-level record.
 
-All multi-byte sequences representing machine words are serialized in
-little-endian format.
+Note that because the tag indicates the length of the value, there is no need
+to use varint encoding of integer values.  Instead, an integer value is encoded
+using the full 8 bits of each byte in little-endian order.  For signed values,
+before encoding, are shifted left one bit, and the sign bit stored as bit 0.
+For negative numbers, the remaining bits are negated so that the upper bytes
+tend to be zero-filled for small integers.
 
-> Note: The `bstring` type is an unusual type representing a hybrid type
-> mixing a UTF-8 string with embedded binary data.  This type is
-> useful in systems like Zeek where data is pulled off the network
-> while expecting a string, but there can be embedded binary data due to
-> bugs, malicious attacks, etc.  It is up to the receiver to determine
-> with out-of-band information or inference whether the data is ultimately
-> arbitrary binary data or a valid UTF-8 string.
+#### 3.2.2 Tag-Encoded Body of Complex Values
 
-A union value is encoded as a container with two elements. The first
-element is the `uvarint` encoding of the index determining the type of
-the value in reference to the union type, and the second element is
-the value encoded according to that type.
+The body of a length-N container comprises zero or more tag-encoded values.
 
 Array, set, and record types are variable length and are encoded
-as a sequence of elements:
+as a sequence of elements and an enum as a small integer:
```

Comment:
```suggestion
as a sequence of elements and enum values as a small integer:
```

The above sentence doesn't mention `map` or `union`.  Rather than adding them, maybe drop the sentence as it's basically paraphrasing the table?